prov/verbs/src/verbs_ep.c

```diff
@@ -121,17 +121,34 @@ void vrb_shutdown_qp_in_err(struct vrb_ep *ep)
 	if (!ep || !ep->ibv_qp || ep->ibv_qp->state == IBV_QPS_UNKNOWN || !ep->eq)
 		return;
 
+	/* Prevent reporting duplicate shutdown events.  These checks are
+	 * racy in the case we hit an error (which is unlikely and unexpected),
+	 * but work in expected situations.  Proper fixes are applied in v1.19.
+	 */
+	ofi_mutex_lock(&ep->eq->lock);
+	if (ep->shutdown)
+		goto unlock;
+	ep->shutdown = true;
+	ofi_mutex_unlock(&ep->eq->lock);
+
 	memset(&attr, 0, sizeof(attr));
 	memset(&init_attr, 0, sizeof(init_attr));
 	if (ibv_query_qp(ep->ibv_qp, &attr, IBV_QP_STATE, &init_attr))
-		return;
+		goto err;
 	if (attr.cur_qp_state != IBV_QPS_ERR)
-		return;
+		goto err;
 
 	memset(&entry, 0, sizeof(entry));
 	entry.fid = &ep->util_ep.ep_fid.fid;
 	if (vrb_eq_write_event(ep->eq, FI_SHUTDOWN, &entry, sizeof(entry)) > 0)
```

Comment:
There is a race in the case of a failure.  I added a comment about that.  The race should be resolved in main/v1.19.x.  For this branch, I'm okay keeping that race because the only time we hit it is if malloc() fails.  If that happens, the app is realistically doomed.

The comments make sense.  I'll update.